python/ray/llm/_internal/serve/deployments/routers/prefix_tree.py

```diff
@@ -0,0 +1,379 @@
+                        curr_node = matched_node
+                        i += shared_count
+            curr_node.tenant_last_access_time[tenant] = timestamp_ms
+            self.tenant_nodes[tenant].add(curr_node)
+            return curr_node
+
+    def prefix_match(
+        self, text: str, available_tenants: Optional[List[str]] = None
+    ) -> Tuple[str, Optional[List[str]]]:
+        """
+        Match text against tree and return (matched_text, matched_tenants).
+        Does not update access time for the matched tenants (only updates when insert() is called).
+        If available_tenants is not provided, all tenants are considered.
+        """
+        if available_tenants:
+            # Filter available_tenants to only include those that exist in the tree
+            available_tenants = [
+                tenant for tenant in available_tenants if tenant in self.tenants
+            ]
+            if not available_tenants:
+                return "", None
+        else:
+            available_tenants = list(self.tenants)
+
+        with self.lock:
+            curr_node: Node = self.root
+            i: int = 0
+            text_len: int = len(text)
+
+            while i < text_len:
+                first_char: str = text[i]
+                curr_text: str = text[i:]
+
+                if first_char in curr_node.children:
+                    matched_node: Node = curr_node.children[first_char]
+
+                    # Check if any of the available tenants match this node
+                    if not any(
+                        tenant in matched_node.tenant_last_access_time
+                        for tenant in available_tenants
+                    ):
+                        break
+
+                    shared_count: int = self.shared_prefix_count(
+                        matched_node.text, curr_text
+                    )
+                    i += shared_count
+                    curr_node = matched_node
+
+                    if shared_count < len(matched_node.text):
+                        # Partial match, stop here
+                        break
+                else:
+                    # No match found, stop here
+                    break
+
+            # Select the tenants in available_tenants that are in the current node
+            selected_tenants: Optional[List[str]] = None
+            matching_tenants = [
+                tenant
+                for tenant in available_tenants
+                if tenant in curr_node.tenant_last_access_time
+            ]
+            if matching_tenants:
+                selected_tenants = matching_tenants
+
+            ret_text: str = text[:i]
+            return ret_text, selected_tenants
+
+    def add_tenant(self, tenant: str) -> None:
+        """Add a tenant to the tree."""
+        with self.lock:
+            if tenant in self.tenants:
+                raise ValueError(f"Cannot add tenant '{tenant}': tenant already exists")
```

Comment:
Do you want to really raise an error here? or is it better to make this idempotent? meaning that if you add the same tenant it basically becomes a no-op (maybe with a warning log)